### ISSUE TITLE
Feature request: ARM builds

### ISSUE THREAD

**Issue by Botspot** [musescore/MuseScore#15251]
Many thousands of people are using ARM linux systems, especially now with Asahi linux running on M1 Macs, in addition to the 46 million Raspberry Pi computers sold.
I am the lead developer for [the world's most popular ARM linux app store](https://github.com/Botspot/pi-apps), and am interested in adding Musescore to it as I have used Musescore v2 in the past on my Pi.

Ideally armhf and arm64 AppImages can be built and included on Github Releases.

Willing to help in any way I can.

**Comment by AntonioBL** [comment]
There is no official support for ARM systems, I think mainly because there are no official Qt5 binaries for ARM from Qt.
I tried an unofficial build of MuseScore 4.0 (using a personal build of Qt5), similar to the unofficial build of MuseScore 3.6.2 co-developed with @MarcSabatella that he hosted on his [website ](https://community.masteringmusescore.com/c/education-resources/install-musescore-on-chromebook).
If you want to test if it works, you can find the Github artifacts [here](https://github.com/AntonioBL/MuseScore/suites/9868241326/artifacts/476111206). But I underline that it is an **unofficial** build.

**Comment by Botspot** [comment]
We have the resources to self-build some apps as necessary. If you could share your process to build QT5 that would be appreciated, as well as any workarounds you found when building the appimages.

In case it helps, I just tried running your armhf appimage and here was the output:
```
$ /home/pi/Downloads/MuseScore-4.0.0.3696976329-armhf.AppImage
/lib/arm-linux-gnueabihf/libjack.so.0
QEventLoop: Cannot be used without QApplication
QEventLoop: Cannot be used without QApplication
11:54:53.115 | INFO  | main_thread | GlobalModule | onPreInit: log path: /home/pi/.local/share/MuseScore/MuseScore4Development/logs/MuseScore_221217_115453.log
11:54:53.116 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.0.0-Development, build number 3696976329 ===
11:54:54.038 | ERROR | main_thread | CrashHandler | start: crash handler not exists, path: /tmp/.mount_MuseSc7e1M2r/bin/crashpad_handler
11:54:54.038 | ERROR | main_thread | DiagnosticsModule | onInit: failed start crash handler
11:54:54.060 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/pi/.local/share/MuseScore/MuseScore4Development/shortcuts.xml"): device not open
11:54:54.062 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/pi/.local/share/MuseScore/MuseScore4Development/midi_mappings.xml"): device not open
11:54:54.247 | ERROR | main_thread | MuseSamplerLibHandler | MuseSamplerLibHandler: Unable to open MuseSampler library, path: libMuseSamplerCoreLib.so
11:54:54.247 | ERROR | main_thread | MuseSamplerResolver | checkLibrary: Incompatible MuseSampler library; ignoring
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failedALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
11:54:55.851 | ERROR | main_thread | WorkspaceManager | findWorkspaceFiles: [401] The file does not exist
11:54:55.851 | WARN  | main_thread | WorkspaceManager | setupDefaultWorkspace: not found default workspace, will be created new
11:54:55.853 | ERROR | main_thread | PluginsConfiguration | updatePluginsConfiguration: [401] The file does not exist
11:54:55.864 | WARN  | main_thread | CloudService | readTokens: Could not find the tokens file: /home/pi/.local/share/MuseScore/MuseScore4Development/cred.dat
11:54:55.865 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
11:54:55.866 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
11:54:55.867 | INFO  | main_thread | IpcSocket  | connect: success connected to ipc server
11:54:55.867 | INFO  | 3593417664 | IpcServer  | listen: id: "6c4e9dae28284322acf9b3faf8eaf2e3" 
11:54:56.213 | WARN  | main_thread | Qt         | QQmlApplicationEngine failed to load component
11:54:56.213 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/platform/linux/Main.qml:30:1: Type AppWindow unavailable

11:54:56.213 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/AppWindow.qml:23:1: plugin cannot be loaded for module "QtQuick.Controls": Cannot protect module QtQuick.Controls 2 as it was never registered

11:54:56.214 | WARN  | main_thread | Qt         | qrc:/qml/platform/linux/Main.qml:30:1: Type AppWindow unavailable
11:54:56.214 | WARN  | main_thread | Qt         | qrc:/qml/AppWindow.qml:23:1: plugin cannot be loaded for module "QtQuick.Controls": Cannot protect module QtQuick.Controls 2 as it was never registered
/tmp/.mount_MuseSc7e1M2r/AppRun: line 26: 28404 Segmentation fault      "${APPDIR}/bin/mscore4portable" "$@"
```

**Comment by Jojo-Schmitz** [comment]
See also https://musescore.org/en/node/306066

**Comment by AntonioBL** [comment]
@Botspot : thank you very much for testing.
I think that the AppImage may be trying to load some Qt libraries from the operating system instead of from its internal library folder.
Can you please try to start the AppImage again, after exporting the following environment variable `export QT_DEBUG_PLUGINS=1`, and attach a log file of the output?

Thank you for your help.

For what concerns Qt5, I used Github Actions to build only the parts needed by MuseScore (3.6): I used a Docker image and had it cross-compile Qt for armhf and arm64.
Here is the Github repository of the build chain (for Qt5.15.2): https://github.com/AntonioBL/qt5DockerMuseScore/tree/qt515
In particular, [this file](https://github.com/AntonioBL/qt5DockerMuseScore/blob/qt515/Recipe1) is the main script, in which I used a couple of workarounds to make Qt and the related MuseScore compilation work (it is not a really elegant file :sweat_smile: ).

Ciao,
ABL

**Comment by MarcSabatella** [comment]
I tried yesterday and ran into issues with QtQuick not being installed.  I started trying to install what I needed manually and got somewhere but still apparently something isn't "registered".  Then I got sidetracked with other issues.  I'll return to this today.

**Comment by MarcSabatella** [comment]
Here is my log when setting that environment variable
[out.txt](https://github.com/musescore/MuseScore/files/10272011/out.txt)

**Comment by AntonioBL** [comment]
Thank you Marc for the test, too.
I think the AppImage was not including the qml files for QtQuick and QtQuick2; in principle it should not be needed to manually install qt modules.
The following artifact should contain those missing files:
https://github.com/AntonioBL/MuseScore/suites/9992337850/artifacts/484978101
and I hope it works now.

**Comment by Botspot** [comment]
> The following artifact should contain those missing files: https://github.com/AntonioBL/MuseScore/suites/9992337850/artifacts/484978101
and I hope it works now.

It got a lot farther this time. The splashscreen appeared for roughly 2 seconds, and then it vanished.
Output below:
```
$ /home/pi/Downloads/MuseScore-4.0.0.3750086940-armhf.AppImage
/lib/arm-linux-gnueabihf/libjack.so.0
QEventLoop: Cannot be used without QApplication
QEventLoop: Cannot be used without QApplication
08:52:16.803 | INFO  | main_thread | GlobalModule | onPreInit: log path: /home/pi/.local/share/MuseScore/MuseScore4Development/logs/MuseScore_221221_085216.log
08:52:16.804 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.0.0-Development, build number 3750086940 ===
08:52:17.881 | ERROR | main_thread | CrashHandler | start: crash handler not exists, path: /tmp/.mount_MuseScAQVQDh/bin/crashpad_handler
08:52:17.881 | ERROR | main_thread | DiagnosticsModule | onInit: failed start crash handler
08:52:17.900 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/pi/.local/share/MuseScore/MuseScore4Development/shortcuts.xml"): device not open
08:52:17.902 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/pi/.local/share/MuseScore/MuseScore4Development/midi_mappings.xml"): device not open
08:52:18.015 | ERROR | main_thread | MuseSamplerLibHandler | MuseSamplerLibHandler: Unable to open MuseSampler library, path: libMuseSamplerCoreLib.so
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
08:52:18.015 | ERROR | main_thread | MuseSamplerResolver | checkLibrary: Incompatible MuseSampler library; ignoring
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
08:52:19.679 | ERROR | main_thread | PluginsConfiguration | updatePluginsConfiguration: [401] The file does not exist
08:52:19.691 | WARN  | main_thread | CloudService | readTokens: Could not find the tokens file: /home/pi/.local/share/MuseScore/MuseScore4Development/cred.dat
08:52:19.692 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
08:52:19.692 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
08:52:19.694 | INFO  | main_thread | IpcSocket  | connect: success connected to ipc server
08:52:19.706 | INFO  | 3589206976 | IpcServer  | listen: id: "674d265c0a8e4399b99ec84c848105dc" 
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: No such file or directory
08:52:20.073 | WARN  | main_thread | Qt         | QQmlApplicationEngine failed to load component
08:52:20.073 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/platform/linux/Main.qml:30:1: Type AppWindow unavailable

08:52:20.073 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/AppWindow.qml:23:1: module "QtQuick.Controls" version 2.15 is not installed

08:52:20.073 | WARN  | main_thread | Qt         | qrc:/qml/platform/linux/Main.qml:30:1: Type AppWindow unavailable
08:52:20.074 | WARN  | main_thread | Qt         | qrc:/qml/AppWindow.qml:23:1: module "QtQuick.Controls" version 2.15 is not installed
/tmp/.mount_MuseScAQVQDh/AppRun: line 26: 22477 Segmentation fault      "${APPDIR}/bin/mscore4portable" "$@"
```

**Comment by AntonioBL** [comment]
@Botspot : Thank you really a lot for your help.
The QtQuick/Controls.2 was still missing, sorry. :sweat_smile: 
Now this should contain (hopefully) all the libraries: https://github.com/AntonioBL/MuseScore/suites/9993912688/artifacts/485101392
Fingers crossed...

**Comment by MarcSabatella** [comment]
Incredible :-).  Thanks, it works (testing aarch64 on my Lenovo Chromebook Duet)!  At least, I can start the programm, but file diaogs don't open.  I get the following:

`10:24:37.575 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/Ui/InteractiveProvider.qml:125: TypeError: Cannot read property 'objectId' of undefined

10:24:37.578 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/Ui/InteractiveProvider.qml:125: TypeError: Cannot read property 'objectId' of undefined
`

Of course, Muse Hub and Muse Sounds re annother matter entirely.

**Comment by MarcSabatella** [comment]
I *can* however create a new score, with basic playback.  But I can't save it since the file dialog never appears.

**Comment by Botspot** [comment]
Same here when running the armhf appimage. Everything works but the score cannot be saved as the file dialog never appears.
This corresponds with the following output on the commandline:
```
11:53:11.297 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/Ui/InteractiveProvider.qml:125: TypeError: Cannot read property 'objectId' of undefined

11:53:11.298 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/Ui/InteractiveProvider.qml:125: TypeError: Cannot read property 'objectId' of undefined
```

**Comment by AntonioBL** [comment]
Thank you both for the tests.
I think it is a bug of MuseScore code, since `objectId` is a direct property of FileDialog, see [this possible fix](https://github.com/AntonioBL/MuseScore/commit/90ac565f737f9c741d7f599526fc73ee5d2b18a2).
And indeed, I had a crash when opening the file menu when I tried to make build using Wayland platform plaugin and this fix seems to solve that problem (but then there are other crashes when clicking the "cancel" button of the File Open dialog that are maybe caused by instabilities of Qt wayland implementation).
Here are the new binaries: https://github.com/AntonioBL/MuseScore/suites/10011638095/artifacts/486380143
:crossed_fingers:

**Comment by Botspot** [comment]
The Save dialog still does not open.
The errors corresponding with this action now are:
```
10:47:14.601 | WARN  | main_thread | Qt         | QObject::connect(QQuickItem, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
10:47:14.601 | WARN  | main_thread | Qt         | QObject::connect(QObject, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
```

**Comment by Botspot** [comment]
Those same errors also occur if I try to close musescore without saving, and click "Don't save". It will not close at all anymore, whereas the last build artifact did.

**Comment by cbjeukendrup** [comment]
Those error messages are probably not related, I'm afraid. They appear for me on macOS too, while everything works fine.

**Comment by AntonioBL** [comment]
I think that the culprit could be a missing GTK+ platform theme that was not included in my compilation of (ARM) Qt: [here](https://doc.qt.io/qt-5/qml-qt-labs-platform-filedialog.html#availability) it says that it is a needed dependency for running the FileDialog QML Type.
I will try to include this platform theme in the compilation in the next days (most probably after Christmas).
Thank you again.
Ciao,
ABL

**Comment by beaver700nh** [comment]
Hello, I come here from #15564 and I am on a Raspberry Pi (Raspbian GNU/Linux 11 (bullseye) armv7l 32-bit). I have also been trying to compile MuseScore 4.0.0. I also compiled Qt by myself. As seen in my issue #15564 it isn't working. So I tried @AntonioBL​'s [AppImages](https://github.com/AntonioBL/MuseScore/suites/10011638095/artifacts/486380143). It kind of works, but as @Botspot mentioned the save dialog isn't working. I also noticed weird black boxes but that may just be due to overclocking.

Here is a sample log output of me trying to save a score:
```
/lib/arm-linux-gnueabihf/libjack.so.0
QEventLoop: Cannot be used without QApplication
QEventLoop: Cannot be used without QApplication
14:04:21.255 | INFO  | main_thread | GlobalModule | onPreInit: log path: /home/pi/.local/share/MuseScore/MuseScore4Development/logs/MuseScore_221224_140421.log
14:04:21.255 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.0.0-Development, build number 3758088051 ===
14:04:21.945 | ERROR | main_thread | CrashHandler | start: crash handler not exists, path: /tmp/.mount_MuseScVhh4Vy/bin/crashpad_handler
14:04:21.945 | ERROR | main_thread | DiagnosticsModule | onInit: failed start crash handler
14:04:21.955 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/pi/.local/share/MuseScore/MuseScore4Development/shortcuts.xml"): device not open
14:04:21.957 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/pi/.local/share/MuseScore/MuseScore4Development/midi_mappings.xml"): device not open
14:04:22.038 | ERROR | main_thread | MuseSamplerLibHandler | MuseSamplerLibHandler: Unable to open MuseSampler library, path: libMuseSamplerCoreLib.so
14:04:22.038 | ERROR | main_thread | MuseSamplerResolver | checkLibrary: Incompatible MuseSampler library; ignoring
14:04:22.779 | ERROR | main_thread | PluginsConfiguration | updatePluginsConfiguration: [401] The file does not exist
14:04:22.788 | WARN  | main_thread | CloudService | readTokens: Could not find the tokens file: /home/pi/.local/share/MuseScore/MuseScore4Development/cred.dat
14:04:22.788 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
14:04:22.788 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
14:04:22.789 | INFO  | main_thread | IpcSocket  | connect: success connected to ipc server
14:04:22.789 | INFO  | 2506683328 | IpcServer  | listen: id: "a5f91d8c8a8e4a7ab4730ea9368ffa35" 
14:04:26.873 | WARN  | main_thread | Qt         | QObject::connect(QQuickItem, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
14:04:26.873 | WARN  | main_thread | Qt         | QObject::connect(QObject, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
14:04:28.477 | ERROR | 2429547456 | UpdateService | parseRelease: failed parse, err: "illegal value" 
14:04:28.478 | ERROR | main_thread | UpdateScenario | doCheckForUpdate: Unable to check for update, error: [1701] 
14:04:30.676 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-new
14:04:30.844 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/Project/internal/NewScore/KeySignatureSettings.qml:116:13: QML KeySignatureListView: Binding loop detected for property "rows"

14:04:30.845 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/Project/internal/NewScore/KeySignatureSettings.qml:116:13: QML KeySignatureListView: Binding loop detected for property "rows"
14:04:31.146 | WARN  | main_thread | Qt         | QObject::connect(QQuickItem, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
14:04:31.146 | WARN  | main_thread | Qt         | QObject::connect(QObject, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
14:04:40.731 | INFO  | main_thread | NotationProject | doSave: success save file: "/home/pi/.local/share/MuseScore/MuseScore4Development/new_project.mscz" 
14:04:41.319 | ERROR | 2548032448 | SynthResolver | resolveSynth: invalid audio source params for trackId: 0
14:04:41.529 | INFO  | 2548032448 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseScVhh4Vy/share/mscore4portable-4.0/sound/MS Basic.sf3
14:04:41.539 | INFO  | 2548032448 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseScVhh4Vy/share/mscore4portable-4.0/sound/MS Basic.sf3
14:04:44.176 | WARN  | main_thread | Qt         | QQmlComponent: Created graphical object was not placed in the graphics scene.
14:04:45.951 | ERROR | 2548032448 | SynthResolver | resolveSynth: invalid audio source params for trackId: 1
14:04:45.967 | INFO  | 2548032448 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseScVhh4Vy/share/mscore4portable-4.0/sound/MS Basic.sf3
14:04:45.979 | INFO  | 2548032448 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseScVhh4Vy/share/mscore4portable-4.0/sound/MS Basic.sf3
14:04:53.030 | WARN  | main_thread | Qt         | QQmlComponent: Created graphical object was not placed in the graphics scene.
14:04:55.381 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
14:04:55.546 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/Project/AskSaveLocationTypeDialog.qml:70: TypeError: Cannot read property 'enabled' of null

14:04:55.546 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/Project/AskSaveLocationTypeDialog.qml:70: TypeError: Cannot read property 'enabled' of null
14:04:55.550 | WARN  | main_thread | Qt         | QObject::connect(QQuickItem, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
14:04:55.550 | WARN  | main_thread | Qt         | QObject::connect(QObject, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
14:04:57.271 | ERROR | main_thread | SaveProjectScenario | askSaveLocation: "ASSERT FAILED!": type != SaveLocationType::Undefined, file: /MuseScore/src/project/internal/saveprojectscenario.cpp, line: 47
/tmp/.mount_MuseScVhh4Vy/AppRun: line 26: 23922 Killed                  "${APPDIR}/bin/mscore4portable" "$@"
```

I had to kill it because it didn't let me close the program.
Hopefully a resolution can be found for this issue.

[image of black boxes](https://media.discordapp.net/attachments/1046460180126249020/1056288346298658826/screenshot_2022-12-24_14-03-12-265981330.png)

**Comment by beaver700nh** [comment]
@AntonioBL​, really appreciate the work you've put in to make these AppImages. May I ask how you did it and got it to work? How did you fix the issues you came across?

**Comment by cbjeukendrup** [comment]
For the black boxes, see #15399.

You could look at Antonio's branch: https://github.com/musescore/MuseScore/compare/master...AntonioBL:MuseScore:4.0_ARM

**Comment by beaver700nh** [comment]
OK will take a look thanks

**Comment by beaver700nh** [comment]
Hmm looks like #15399 has no resolution as of yet

**Comment by theofficialgman** [comment]
I see the main issue happening here has to do with qt5 compilation. You are welcome to follow the process I use for building prism launcher on bionic with qt5.15.2 ppa. Don't try to actually use the ppa on buster, it won't work, but you can follow the process of building and then copying the libs over into your package (appimage or Deb) at the end

https://le-builds.lakka.tv/Switch.aarch64/?C=M&O=D

**Comment by beaver700nh** [comment]
Sorry... you kinda lost me there

**Comment by AntonioBL** [comment]
Ciao.
@beaver700nh :
I managed to cross-compile Qt for ARM inside a docker image, cross-compile MuseScore inside the same image, and then used an ARM docker image to package MuseScore. All of this automatically done by Github actions. I wanted to do everything inside an ARM docker image, but "native" Qt compilation was taking too long and the job was cancelled by Github.
The missing dialog is due to the fact that the GTK+ platform theme was not compiled. I am trying to find all the needed dependencies so that it is included in the compilation I am using (unfortunately, Qt dependencies for the comppilation of the different modules are not clearly listed somwhere, or at least I couldn't find the place where they are listed).
As @cbjeukendrup pointed out, you can find my (messy) work in these repositories:
- Qt cross-compilation (inside a docker image): https://github.com/AntonioBL/qt5DockerMuseScore/tree/qt515
- ARM docker image for packaging: https://github.com/AntonioBL/armpackagingMuseScore/tree/qt5.15.2
- MuseScore patched for ARM (cross-)compilation and packaging: https://github.com/AntonioBL/MuseScore/tree/v4.0arm

**Comment by MarcSabatella** [comment]
@AntonioBL - prior to https://github.com/musescore/MuseScore/pull/13847, MuseScore would use whichever platform theme Qt defaulted to.  This was changed to force use GTK3 here in this PR, in order to fix some accessibility and other issues with other themes.  The PR made other changes, but I wonder if merely reverting this portion:

https://github.com/musescore/MuseScore/pull/13847/files#diff-ae8bd386aa34e80f31820a26f22d109e01ea8a17d4521a2908dbf88f68a88c9aR71-R74

or tightening the ifdef to exclude ARM builds would allow it to work using other platform themes?  I had already tested your build with other settings of QT_QPA_PLATFORMTHEME but it had no effect (not that I really expected it to) because it's always being forced to GTK3.

The default platform theme on my other systems was not ideal in that keyboard navigation didn't function correctly within these dialogs, but I'd trade that for a working build :-).

**Comment by beaver700nh** [comment]
I have been able to get the save dialog to appear very briefly but then the program segfaults. See #15564

**Comment by AntonioBL** [comment]
I think I may have found a way to compile the GTK+ platform theme. I am now waiting for the server to (hopefully) finish the compilation... :crossed_fingers:

**Comment by beaver700nh** [comment]
Hey @AntonioBL any luck with that compilation?

**Comment by AntonioBL** [comment]
Ciao.
I had a couple of problems with uploading the Docker image: I tried to update the Github action, but they had changed the input parameters and split it into two distinct actions, so in the end I reverted to the old Docker build-and-push action. 😅 
Here is the latest build:
https://github.com/AntonioBL/MuseScore/suites/10097820372/artifacts/492292693

I will check if other libraries are missing wrt to official release, but libqgtk3.so now is present in the Github artifact.

**Comment by beaver700nh** [comment]
Hello, it seems that now the first save dialog ("Save to cloud or locally?") does not appear either.
```
06:48:05.449 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/Project/internal/NewScore/KeySignatureSettings.qml:116:13: QML KeySignatureListView: Binding loop detected for property "rows"

06:48:05.449 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/Project/internal/NewScore/KeySignatureSettings.qml:116:13: QML KeySignatureListView: Binding loop detected for property "rows"
06:48:05.714 | WARN  | main_thread | Qt         | QObject::connect(QQuickItem, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
06:48:05.715 | WARN  | main_thread | Qt         | QObject::connect(QObject, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
06:48:11.145 | INFO  | main_thread | NotationProject | doSave: success save file: "/home/pi/.local/share/MuseScore/MuseScore4Development/new_project.mscz" 
06:48:11.816 | ERROR | 2510283712 | SynthResolver | resolveSynth: invalid audio source params for trackId: 0
06:48:12.090 | INFO  | 2510283712 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseSc3V1c9y/share/mscore4portable-4.0/sound/MS Basic.sf3
06:48:12.102 | INFO  | 2510283712 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseSc3V1c9y/share/mscore4portable-4.0/sound/MS Basic.sf3
06:48:15.386 | WARN  | main_thread | Qt         | QQmlComponent: Created graphical object was not placed in the graphics scene.
06:48:16.610 | ERROR | 2510283712 | SynthResolver | resolveSynth: invalid audio source params for trackId: 1
06:48:16.627 | INFO  | 2510283712 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseSc3V1c9y/share/mscore4portable-4.0/sound/MS Basic.sf3
06:48:16.643 | INFO  | 2510283712 | FluidSynth | addSoundFonts: success load soundfont: /tmp/.mount_MuseSc3V1c9y/share/mscore4portable-4.0/sound/MS Basic.sf3
06:48:17.666 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
06:48:32.050 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: pad-note-4
06:48:33.624 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: put-note
06:48:35.577 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: note-input-steptime
06:48:37.012 | WARN  | main_thread | Qt         | QQmlComponent: Created graphical object was not placed in the graphics scene.
06:48:38.609 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
06:48:38.611 | WARN  | main_thread | Qt         | QEventLoop::exec: instance 0x44f4094 has already called exec()
06:48:45.808 | WARN  | main_thread | Qt         | QQmlComponent: Created graphical object was not placed in the graphics scene.
06:48:47.228 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
06:48:47.234 | WARN  | main_thread | Qt         | QEventLoop::exec: instance 0x44f4094 has already called exec()
06:48:48.395 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
06:48:48.396 | WARN  | main_thread | Qt         | QEventLoop::exec: instance 0x44f4094 has already called exec()
06:48:48.644 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
06:48:48.646 | WARN  | main_thread | Qt         | QEventLoop::exec: instance 0x44f4094 has already called exec()
06:48:48.835 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: file-save
06:48:48.836 | WARN  | main_thread | Qt         | QEventLoop::exec: instance 0x44f4094 has already called exec()
```

**Comment by AntonioBL** [comment]
😞  Something is probably still missing.
I run a comparison between an official AppImage and this artifact and I found that these files/folders are missing:
1)
`fallback/libnss3.so/`
2)
`qml/Qt/labs/platform/libqtlabsplatformplugin.so`
`qml/Qt/labs/platform/plugins.qmltypes`
`qml/Qt/labs/platform/qmldir`
3) the compiled qml files (".qmlc") in these folders (the ".qml" counterpart is present):
`qml/QtGraphicalEffects/private/`
`qml/QtQuick/Controls/`
`qml/QtQuick/Dialogs/`
`qml/QtQuick/Extras/`
4) the translation files (".qm") in this folder:
`share/mscore4portable-4.0/locale/`

I think the remaining issue preventing the file dialog from appearing is 2).
1) is just a fallback folder and it should not be a problem. I think also 3) should not be a problem (maybe), since the qml files are present. 4) probably just needs an additional build step that I missed.

At the moment I don't know which dependency I am missing for Qt.labs.platform.plugins.

**Comment by beaver700nh** [comment]
Hey @AntonioBL,
From my attempt to compile Qt 5.15.7 by myself, I have just the three files needed for issue `3)`. See here:
[platform.zip](https://github.com/musescore/MuseScore/files/10320131/platform.zip)

**Comment by AntonioBL** [comment]
I found that they are also present in the Qt5 compilation done via Github actions, I tried to force their installation, but now I think something was updated on the server side and MuseScore does not cross-compile anymore (ALSA not found when crosscompiling). Working on it...

**Comment by beaver700nh** [comment]
Okay. In case it helps, I have also found these files in the preexisting installation of Qt5.
[platform_pre.zip](https://github.com/musescore/MuseScore/files/10320287/platform_pre.zip)

I might also be able to compile for you, as I am currently on a Raspberry Pi (Raspberry Pi 4 Model B Rev 1.4).

**Comment by theofficialgman** [comment]
> Many thousands of people are using ARM linux systems, especially now with Asahi linux running on M1 Macs, in addition to the 46 million Raspberry Pi computers sold.

I'd just like to add there are also now 115+ million Nintendo Switch consoles capable of running Linux. Far outnumbering the pi and apple silicon macs. 😉

**Comment by AntonioBL** [comment]
... I had done a really silly typo...
Here is the latest build, comprising also the translations:
https://github.com/AntonioBL/MuseScore/suites/10101866500/artifacts/492548862
🤞 🤞 🤞

**Comment by beaver700nh** [comment]
Video: https://www.dropfiles.pro/jiZKpB1jY7LzTra

**Comment by AntonioBL** [comment]
Sorry, I have no clue about what is happening. All the dependencies now should be present, so I think it could be a bug in MuseScore file dialog implementation, or in Qt itself when working in an ARM system.
I tried to follow Marc's suggestion from [this comment](https://github.com/musescore/MuseScore/issues/15251#issuecomment-1365949876), but I think the file dialog should also somehow be changed.
Here are the binaries: https://github.com/AntonioBL/MuseScore/suites/10114196789/artifacts/493404769
Unfortunately, I do not have an ARM system, and I don't know how to emulate it (I once tried with Qemu; it somehow worked, but it was very difficult to properly set the virtual machine, and then it was extrememly slow).

**Comment by beaver700nh** [comment]
Okay, well thank you for your time in attempting to get it working.

**Comment by MarcSabatella** [comment]
When I try the version that doesn't force GTK3, I get the Qt save dialog, but even after typing a name and hittig Save, I get the following:

10:09:27.698 | ERROR | main_thread | SaveProjectScenario | askSaveLocation: "ASSERT FAILED!": type != SaveLocationType::Undefined, file: /MuseScore/src/project/internal/saveprojectscenario.cpp, line: 47

followed by the same message as shown about above actually saving to a totally different name than I gavcve:

0:10:39.423 | INFO  | main_thread | NotationProject | doSave: success save file: "/home/marcsabatella/.local/share/MuseScore/MuseScore4Development/new_project.mscz" 

If I force QT_QPA_PLATFORMTHEME=gtk3 myself, save brings up the gtk3 diakog, but when I enter a name and hit Save, the dialog clsoes but the score isn't saved at all.  A subsequent save does report it saved to the new_project.mscz in .local - not the name or folder I specified (which was just test.mscz in the default folder)

So close indeed!

**Comment by MarcSabatella** [comment]
Oh, also: quit doesn't work.  I get the diaog asking if I want to save, I say don't save, the dialog closes, and MuseScore stays up.  Console says:

10:31:39.845 | INFO  | main_thread | ActionsDispatcher | dispatch: try call action: quit
10:31:39.911 | WARN  | main_thread | Qt         | QObject::connect(QQuickItem, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter
10:31:39.911 | WARN  | main_thread | Qt         | QObject::connect(QObject, mu::uicomponents::PopupViewCloseController): invalid nullptr parameter

**Comment by beaver700nh** [comment]
I have exactly the same problems as both of your comments!

**Comment by Botspot** [comment]
> Unfortunately, I do not have an ARM system, and I don't know how to emulate it (I once tried with Qemu; it somehow worked, but it was very difficult to properly set the virtual machine, and then it was extrememly slow).

@AntonioBL, would you like to remote into my Pi4? I've given other devs access in the past to compile and test things on an ARM system.
You can connect with Anydesk, Teamviewer, or Rustdesk. Just send me an email at:
```
echo ai5hbGxlbi5tYXJjaGFuZEBvdXRsb29rLmNvbQo= | base64 -d
```

**Comment by beaver700nh** [comment]
Okey, so I have succeeded in installing MuseScore from the Arch repo even though I am running Debian.
So janky but it works good enough and I'm tired.

**Comment by Botspot** [comment]
Any further comments, @AntonioBL? Have you seen my previous comment on this issue from 5 days ago?

**Comment by AntonioBL** [comment]
@Botspot : thank you for your kind offer! But I think I managed to do on my machine.
Sorry for not answering before: I was busy with our concert band on the first two days of January.
And then I spent a coupe of days trying to run MuseScore for ARM inside qemu. I finally managed to do it, even if it runs extremely slowly (and the audio doesn't work), but it made me possible to cross-compile and check it on the same machine.
Indeed, it is easier for me to debug if I stay on the same machine, without having to wait for the full recompilation on Github servers.
After some investigations, I found that I had made a modification I shouldn't have made, that caused the bug. 😅 
Now it should work. Here are the ARM AppImages:
https://github.com/AntonioBL/MuseScore/suites/10224165195/artifacts/501064959
I hope there are no additional (besides the ones already present in the released version 😉 ) bugs left.

I think the remaining item is to compile it as a "stable" version instead of a Development one. If these binaries work, I will try that (and maybe I will try to clean up the commits I added to the ARM branch).
Ciao and thank you for your patience.

**Comment by MarcSabatella** [comment]
The aarch64 version is now working for me on my Lenovo Chromebook Duet - save & open dialogs and audio included!  I haven't done any extensive testing, but so far so good.

Thanks again for your hard work @AntonioBL .  I guess the next question is, how automatable can this be made with respect to future updates and with respect to allowing others to kick off the build?  We kind of got lucky that 3.6.2 got no updates for almost two years (!) so the last ARM build you made was good for a long time.  But I expect now we'll see 4.x updates much more often.  It would be nice for any of us to be able to build a stable version at any time.

**Comment by Botspot** [comment]
> The aarch64 version is now working for me on my Lenovo Chromebook Duet - save & open dialogs and audio included! I haven't done any extensive testing, but so far so good.

Same here. I will add the current appimages to [Pi-Apps](https://github.com/Botspot/pi-apps), with a note saying it's unstable. That will bring more testers and I'll direct any issues here.

**Comment by theofficialgman** [comment]
Any chance you could build QT5 on bionic @MarcSabatella instead of whatever you built it on?
the appimage you built is requiring glibc 2.28+
```
./MuseScore-4.0.0.3856641905-aarch64.AppImage 
/usr/lib/aarch64-linux-gnu/libjack.so.0
/tmp/.mount_MuseScnPjvSV/bin/mscore4portable: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /tmp/.mount_MuseScnPjvSV/bin/../lib/libQt5Core.so.5)
```

**Comment by theofficialgman** [comment]
I generally use launchpad whenever possible to produce builds. you are welcome to use my qt5.15.2 armhf/arm64 bionic ppa if it is suitable https://launchpad.net/~theofficialgman/+archive/ubuntu/opt-qt-5.15.2-bionic-arm

**Comment by AntonioBL** [comment]
@MarcSabatella : thank you for the test!
@theofficialgman : thank you for pointing me to your Qt ppa. The main struggle I had with MuseScore compilation for ARM was how to find Qt 5.15 binaries for an old enough version of Linux for ARM.
By using this ppa, we can try to build MuseScore completely in an emulated ARM machine, by using qemu and ARM docker images.
Sorry for the silence from my side this week, but I had very little time to work on this compilation (mainly solving the problem of missing dependencies or finding the proper version to build for the AppImage depolyment tools) and I was not sure how much it could take me. But now I think I am very close to have an armhf and arm64 version compiled in Ubuntu Bionic Beaver.

**Comment by MarcSabatella** [comment]
Thanks for the update!  Speaking of which, I was just about to come here to report that it looks like 4.0.1 is imminent.  Fixes some serious corruption & crash issues, so probably worth waiting on that to become official (not sure if that's actually today or just very soon) and then based an initial "release" build on that branch.

**Comment by theofficialgman** [comment]
@AntonioBL I saw your latest actions was having issues related to packaging the appimage due to mutiple architecture in the appdir foder so I forked your repo and adding some logging to figure it out

https://github.com/theofficialgman/MuseScore/actions/runs/3921279696/jobs/6703482660#step:6:17373

the issue is there is literally ONE x86-64 binary that got added to the appdir and everything else is aarch64

```
MuseScore-4.0.0-aarch64.AppDir/bin/crashpad_handler:                                                                                                 ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
```

**Comment by theofficialgman** [comment]
and here is the culprit @AntonioBL , MuseScore is bad and put a binary directly in the repo <https://github.com/musescore/MuseScore/blob/master/src/diagnostics/crashpad_handler/linux/x86-64/crashpad_handler>

CC @cbjeukendrup since this appear to be your binary. is this closed source or can it be buit for arm64 and armhf linux as well please?

**Comment by cbjeukendrup** [comment]
It's not mine, I'm apparently just the last one who touched that file :)

I have no idea from which version of `crashpad_handler` this is; @igorkorsukov may know.

The reason that it is in the repo as a precompiled binary is:
- it's annoying if the same executable needs to be compiled again every single time (also on CI, where we do clean builds)
- it uses a custom build system while we use CMake; of course we can write CMake files for it, but having to do that is not great

However, you can compile the executable yourself in whatever way, and then specify the path to your version using the `CRASHPAD_HANDLER_PATH` CMake option.

(And unfortunately, my recent fixes related to this were not ported to the 4.0.1 branch, so you'll have to use the master branch.)

**Comment by theofficialgman** [comment]
its fine, I went ahead and cherry picked those two commits to a new branch based on 4.0.1 release. shoud be fine for now

**Comment by theofficialgman** [comment]
> * it uses a custom build system while we use CMake; of course we can write CMake files for it, but having to do that is not great
> 
> However, you can compile the executable yourself in whatever way, and then specify the path to your version using the `CRASHPAD_HANDLER_PATH` CMake option.

yeah these do kinda suck to compile. I ran into two problems. one detailed here by another user: https://groups.google.com/a/chromium.org/g/crashpad-dev/c/iLoeoOyt8B4 and another with it not liking to build with clang-6 which is the bionic defaut. luckily installing clang-10 from the bionic repo and adding that to temporarily to path (so it would be found first) works without having to change the nasty buildsystem `PATH=/usr/lib/llvm-10/bin:$PATH`

**Comment by theofficialgman** [comment]
[crashpad-arm64-linux.tar.gz](https://github.com/musescore/MuseScore/files/10420891/crashpad-arm64-linux.tar.gz)

here are the crashpad binaries for arm64 linux for anyone interested (built off the latest commit on main)

**Comment by igorkorsukov** [comment]
> 

We use the crashpad from here (there are some edits for linux) https://github.com/musescore/crashpad_fork

**Comment by theofficialgman** [comment]
ugh... I picked the commits and added the cmake arg but it still installed the x86_64 binary
```
MuseScore-4.0.1-armv7l.AppDir/bin/crashpad_handler:                                                                                                 ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
```

**Comment by theofficialgman** [comment]
@cbjeukendrup this doesn't work still it seems
https://github.com/theofficialgman/MuseScore/commit/a6f18f70802f601fe092c220a20519bb7d459528

I have the commit in my branch and crashpad disabled https://github.com/theofficialgman/MuseScore/commit/bbcd90b3666a807b6f0896085901372312ac296d yet it still installed it.

**Comment by theofficialgman** [comment]
oh nevermind... I seem to have misunderstood the function of the variables
I thought you wrote ninja_build.sh to take those variables and write them directly as cmake args, that is not the case

**Comment by theofficialgman** [comment]
@AntonioBL the builds finshed but unfortunatly segfault on startup

the 4.0.1 branch is here https://github.com/theofficialgman/MuseScore
artifacts are here https://github.com/theofficialgman/MuseScore/actions/runs/3925970087

here is the output
```
/usr/lib/aarch64-linux-gnu/libjack.so.0
/usr/lib/aarch64-linux-gnu/libnss3.so
QEventLoop: Cannot be used without QApplication
QEventLoop: Cannot be used without QApplication
22:59:04.044 | INFO  | main_thread | GlobalModule | onPreInit: log path: /home/garrett/.local/share/MuseScore/MuseScore4/logs/MuseScore_230115_225904.log
22:59:04.044 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.0.1, build number 230160014 ===
22:59:04.278 | WARN  | main_thread | DiagnosticsModule | onInit: crash handling disabled
22:59:04.285 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/garrett/.local/share/MuseScore/MuseScore4/shortcuts.xml"): device not open
22:59:04.286 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/garrett/.local/share/MuseScore/MuseScore4/midi_mappings.xml"): device not open
22:59:04.360 | ERROR | main_thread | MuseSamplerLibHandler | MuseSamplerLibHandler: Unable to open MuseSampler library, path: libMuseSamplerCoreLib.so
22:59:04.360 | ERROR | main_thread | MuseSamplerResolver | checkLibrary: Incompatible MuseSampler library; ignoring
22:59:05.751 | ERROR | main_thread | PluginsConfiguration | updatePluginsConfiguration: [401] The file does not exist
22:59:05.752 | WARN  | main_thread | CloudService | readTokens: Could not find the tokens file: /home/garrett/.local/share/MuseScore/MuseScore4/cred.dat
22:59:05.753 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
22:59:05.753 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
22:59:05.754 | INFO  | main_thread | IpcSocket  | connect: success connected to ipc server
22:59:05.756 | INFO  | 546908754128 | IpcServer  | listen: id: "00c0cb9ff51045329d70732727b2bcb8" 
22:59:06.872 | WARN  | main_thread | Qt         | QQmlApplicationEngine failed to load component
22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/platform/linux/Main.qml:30:1: Type AppWindow unavailable

22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/AppWindow.qml:68:5: Type ToolTipProvider unavailable

22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/Ui/ToolTipProvider.qml:41:26: Type StyledToolTip unavailable

22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/UiComponents/StyledToolTip.qml:28:1: Type StyledPopupView unavailable

22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/UiComponents/StyledPopupView.qml:63:18: Type PopupContent unavailable

22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/UiComponents/internal/PopupContent.qml:97:9: Type StyledDropShadow unavailable

22:59:06.872 | ERROR | main_thread | AppShell   | run: error: qrc:/qml/MuseScore/UiComponents/StyledDropShadow.qml:23:1: module "QtGraphicalEffects" is not installed

22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/platform/linux/Main.qml:30:1: Type AppWindow unavailable
22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/AppWindow.qml:68:5: Type ToolTipProvider unavailable
22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/Ui/ToolTipProvider.qml:41:26: Type StyledToolTip unavailable
22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/UiComponents/StyledToolTip.qml:28:1: Type StyledPopupView unavailable
22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/UiComponents/StyledPopupView.qml:63:18: Type PopupContent unavailable
22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/UiComponents/internal/PopupContent.qml:97:9: Type StyledDropShadow unavailable
22:59:06.873 | WARN  | main_thread | Qt         | qrc:/qml/MuseScore/UiComponents/StyledDropShadow.qml:23:1: module "QtGraphicalEffects" is not installed
./AppRun: line 26: 20886 Segmentation fault      (core dumped) "${APPDIR}/bin/mscore4portable" "$@"
```

**Comment by theofficialgman** [comment]
well that was a pretty easy issue to solve
you forgot to install this package `qt515graphicaleffects`.
I extraced the appimage and added that manually to test and now everything works!

hopefully linuxdeploy picks that up automatically when I add it

**Comment by AntonioBL** [comment]
@theofficialgman : thank you for your great work!
I had no time on Sunday to look at the builds.
I checked the artifact from https://github.com/theofficialgman/MuseScore/actions/runs/3927119883
Comparing to the official (x86_64) AppImage, these files are still missing:
libcrypto
libhogweed
libQt5RemoteObjects
libssl
libxcb-glx
libXext
plugins/imageformats/libqicns
plugins/imageformats/libqtga
plugins/imageformats/libqtiff
plugins/imageformats/libqwbmp
plugins/imageformats/libqwebp
plugins/xcbglintegrations/libqxcb-glx-integration
qml/RemoteObjects/libqtqmlremoteobjects
qml/RemoteObjects/plugins.qmltypes
qml/RemoteObjects/qmldir
All the translations

I think that Qt5RemoteObjects is still missing from the installation script. I wonder why the translations were not included in the artifact. I thought the scripts were launching the generation of the translation files, but maybe I missed one step.

I don't know if the missing llibs in plugin/imageformats are really needed; maybe it is simply a difference with respect to "using OS image libraries" and "using Qt image libraries".

**Comment by AntonioBL** [comment]
Yes, I think I missed `qt515translations` and `qt515remoteobjects`, sorry.

**Comment by theofficialgman** [comment]
thanks for that list.
I went through the official appimage myself as well as mine as well and found libxcb-glx present in both (the qt plugin and the host binary)

libcrypto
libhogweed
libssl
these are all related to qtssl network auth. bionic/focal still have openssl1 so its not a problem not to have it for them, but jammy has openssl3 only so you would get errors if it ever tried to be used. other projects I have seen manually add them to the appimage. I'm not sure how the official appimage already has them. They aren't dynamically linked and the qt5152 .7z they use doesn't even have them. they aren't in the excludelist either. I will leave these missing for now and see if they magically appear when the final packages you noted above are added

libXext is probably just a difference between the qt5 builds they are using and mine. nothing seems to dynamically link to it so it isn't included

**Comment by AntonioBL** [comment]
@theofficialgman: Thank you.
In the meantime, I had launched a build based on your branch (v4.0arm_bionic) where I added those Qt dependencies:
https://github.com/AntonioBL/MuseScore/actions/runs/3929513490
It just ended, but I found that linuxdeploy did not include the libraries from `qt515remoteobjects`.
Maybe those libraries are not really needed...

libssl3.so is present in the `fallback\libnss3.so` directory inside the AppImage, so maybe there will not be a crash when using it in Ubuntu Jammy...

**Comment by theofficialgman** [comment]
also, linuxdeploy-qt chooses what it thinks might be needed for the package during install. it doesn't copy all qt libraries if they aren't needed. the qt5152.7z file that the x86_64 scripts use have basically all QT plugins included but the resulting appimage only has what is needed.

so we could just install all packages from my ppa and then let linuxdeploy-qt choose on its own what is relavent.

**Comment by theofficialgman** [comment]
> linuxdeploy did not include the libraries from `qt515remoteobjects`. Maybe those libraries are not really needed...

I don't see `qml/RemoteObjects` in the official musescore 4.0.1 appimage. where did you see it?

**Comment by theofficialgman** [comment]
> there will be a switch to Qt6 at some point of the release history of MuseScore 4.x (and that's why some OS versions, such as Windows 32bits, or older MacOS's, not supported by Qt6 were already dropped in MuseScore 4.0).

whenever qt6 drops we can rework this again. I have (another persons) qt6 ppa that I use

**Comment by AntonioBL** [comment]
> > Thank you for your hard work on this. For MuseScore 4.0.1 and following, we can try to cherry-pick these commits and launch a build with Github actions.
> 
> I have aleady been building a 4.0.0 and 4.0.1 branch since yesterday this is a 4.0.1 build https://github.com/theofficialgman/MuseScore/actions/runs/3931411531

I know :wink: : this morning I have rebased my 4.0arm_bionic branch onto yours, and I have seen also the other one. Thank you again!
MuseScore 4.0.1 has not been officially released yet, so I was just waiting for its official release (and the official ref commit for 4.0.1).

**Comment by AntonioBL** [comment]
> > I know wink : this morning I have rebased my 4.0arm_bionic branch onto yours, and I have seen also the other one. Thank you again! MuseScore 4.0.1 has not been officially released yet, so I was just waiting for its official release (and the official ref commit for 4.0.1).
> 
> it released 3 days ago. https://github.com/musescore/MuseScore/releases/tag/v4.0.1 I used this as the reference commit

Ah, ok: I missed the announcement :sweat_smile:
So we can use the artifacts of https://github.com/theofficialgman/MuseScore/actions/runs/3931411531 as the unofficial "releases" of MuseScore 4.0.1 for arm systems.

**Comment by MarcSabatella** [comment]
For the record, there was no actual announcement beyond the update message within MuseScore or Muse Hub, plus the fact that there was the release created on GitHub which triggers email announcements if you're subscribed.

Anyhow, thanks everyone for the work on this, looking forward to testing the build when it completes :-).  Assuming all is well, does anyone have thoughts on the best semi-permanent
place to host these builds going forward?  I just used my own Google Drive account for 3.6.2 but maybe there is a way to make the build artifacts more permanent on someone's GitHub account?  I don't know how that is done really, but it does seem like that is where the official 4.0.1 builds live.

**Comment by theofficialgman** [comment]
the builds will live in our files repo along with other pi-apps ports to arm https://github.com/Pi-Apps-Coders/files/releases/tag/large-files

I would ideally like to upstream this to MuseScore rather than carry the patches. they are quite minor https://github.com/musescore/MuseScore/compare/4.0.1...theofficialgman:MuseScore:v4.0.1arm_bionic?expand=1 with the only thing that really needs to be done now is add the `osuosl/publish.sh`  step but I think upstream can do that easily if they want it.

**Comment by MarcSabatella** [comment]
I grabbed the artifact from build 14, and it working great on my arm64 Chromebook.  One minor glitch - on startup, it says there is an update available, and if I choose to install it, it just grabs the x86_64 version.  Not sure if there is some way to maybe disable, or better yet reroute, the check for updates.  But not a big deal at all, this is just fantastic to have it working at all.

Let me know when there are specific URL's I can add to my own unofficial Chromebook installation documentation.

**Comment by theofficialgman** [comment]
yeah I'm uploading the builds now. slow internet here

I "think" that in latest builds musescore should be only checking for armv7l and aarch64 new versions. not x86_64
build # 13 is actually 4.0.1, # 14 is 4.0.0. 

I thought that this commit of mine would fix it checking for x86_64 appimages but maybe there is more too it:
https://github.com/theofficialgman/MuseScore/commit/6fa0f97f47b1cf157308c5db56d8ec2d08b66662

**Comment by theofficialgman** [comment]
> better yet reroute, the check for updates.

I was hoping to get this included in musescore itself, so no re-routing of updates
though it shouldn't be hard to do that. it should just be a matter of changing the update_information variable to point to a new repo or web url https://github.com/theofficialgman/MuseScore/blob/6fa0f97f47b1cf157308c5db56d8ec2d08b66662/build/ci/linux/package.sh#L90

**Comment by theofficialgman** [comment]
I think this file is to blame for the misbehaving update check
https://github.com/musescore/MuseScore/blob/master/src/update/internal/updateservice.cpp

@Eism can you explain what this new update functionality does and how it works with the previously existing zsync https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information implementation for linux? it looks to me like maybe you don't use the zsync implemenation anymore

**Comment by cbjeukendrup** [comment]
I believe there are two update systems on Linux; one is built-in in the app itself (that one uses `updateservice.cpp`), and one is provided by the AppImage by running something like `MuseScore.AppImage update` (and that one uses the zsync files).

**Comment by theofficialgman** [comment]
ah ok, so its the new update functionality that I pointed out that is problematic. its only checking the filetype (appimage) and nothing else. so when there is an x86_64 appimage from github it downloads that. I will disable it from here on

**Comment by MarcSabatella** [comment]
This sounds like very good news, but can someone explain what this means in practice?  There won't be ARM AppImages created automatically as part of ordinary PR or nightly builds, but they can be created on demand for releases, and may actually be made available through the normal channels (eg, hosted on GitHub?)

**Comment by theofficialgman** [comment]
@MarcSabatella the actions are not currently set to run on a schedule for these arm builds. However, it is fully setup to do a manual workflow trigger which will automatically upload the builds to `osuosl` server (https://ftp.osuosl.org/pub/musescore-nightlies/linux/4x/nightly/?C=M;O=D), sentry for the symbols dump, and of course also artifacts in the actions log that can be made into a release.

you, at musescore, can enable the scheduled builds if you want easily by adding them back to the `on` in the yml, but I felt like that was unnecessary since they take so long to complete
```

**Comment by theofficialgman** [comment]
I would recommend that someone with the privledges go ahead and run a nightly build through the actions tab to test that those systems are functional
https://github.com/musescore/MuseScore/actions/workflows/ci_linux_arm_mu4.yml

**Comment by theofficialgman** [comment]
> 1. For users who want an ARM AppImage for 4.0.1, they should get it from the Pi-Apps-Coders repository at https://github.com/Pi-Apps-Coders/files/releases/tag/large-files

thats correct. however it is an unofficial build missing the crashpad handler, musescore api keys for youtube (even though I think they are unused currently)

> 2. Nothing about the merge of the PR into master changes this

the merge into master doesn't affect previously released 4.0.1 builds

> 3. If your PR is merged into 4.0.2 and a manual build of the new action is performed, the artifact will be uploaded to the MuseScore GitHub account alongside the other official builds, so I will be able to direct users there instead of to the Pi-Apps-Coders repository

publishing on musescore github is done manually following this process <https://github.com/musescore/MuseScore/wiki/Release-process#prepare-builds> . artifacts from github actions are never uploaded directly to a github release. A maintainer manually runs the actions in release mode for each target platform, downloads the artifacts, and then creates a tag and release following that wiki page process I linked.

> 4. If the PR is _not_ merged into 4.0.2, or if it is merged but a manual build is not performed, then we will need to continue relying on unofficial builds in the Pi-Apps-Coders repository.  But the manual build could in theory be run at any time.

that is correct

> 5. Until there is an official build, it will remain possible for you to update Pi-Apps-Coders, and the "update" command line option will work there, but the fancy new in-app update will not

no. the update command line option is not present in the current Pi-Apps-Coders build (I made the builds before this was implemented https://github.com/musescore/MuseScore/pull/15923/commits/7c4e67bea119e91e62088d8285af067eb3d01997 ) and the in app updater does not automatically update musescore on linux anyway regardless. On linux, the in app updater just opens a folder with a new appimage downloaded from the musescore server. the official builds from this PR have the in app updater disabled (since it is still broken: https://github.com/musescore/MuseScore/issues/15251#issuecomment-1384543079 ) and only the update command line option implemented.

> 6. Once an official has an AppImage is made available, then the command line "update" process will work normally

yes

> as will the fancy new in-app update process

no, see above. the in-app updater is disabled https://github.com/musescore/MuseScore/blob/master/build/ci/linux/build.sh#L72-L75

**Comment by MarcSabatella** [comment]
OK, thanks, that helped a lot!  So it seems the PR makes it *possible* for the maintainers to kick off a manual build (for master now, for 4.0.2 if this and other relevant PR's are ported there), then manually add the artifacts to the release, but it still remains to be seen if/when this happens.
